GC sweeping should not be done in the End phase.

MenloDorian opened the pull request:
#### b5808ff792b03d7cfd54a963340fdbb0ab4ec1b9
<pre>
GC sweeping should not be done in the End phase.
<a href="https://bugs.webkit.org/show_bug.cgi?id=303634">https://bugs.webkit.org/show_bug.cgi?id=303634</a>
<a href="https://rdar.apple.com/165681233">rdar://165681233</a>

Reviewed by Dan Hecht and Keith Miller.

That&apos;s because the end phase may be run on the GC thread instead of the mutator thread.
Sweeping needs to be done on the mutator thread because AtomStringImpl destruction needs
to unregistered from the mutator&apos;s thread local AtomStringTable.  This can only be done
on the mutator thread.

303051@main introduced eager synchronous sweeping in the GC End phase if there&apos;s a critical
memory condition.  We&apos;ll fix this by changing this to do the critical memory check in
Heap::shouldSweepSynchronously() instead, and simply let Heap::finalize() do the sweeping
in the mutator thread instead (which it can already do).

Added a RELEASE_ASSERT in Heap::sweepSynchronously() to verify that the thread is holding
the API lock (as a proxy for identifying that it is the mutator thread).  This should
help catch this issue with existing tests to the extent possible.  The bug does rely on
a race condition that is not easily reproducible.

* Source/JavaScriptCore/heap/Heap.cpp:
(JSC::Heap::sweepSynchronously):
(JSC::Heap::runEndPhase):
(JSC::Heap::shouldSweepSynchronously):
* Source/JavaScriptCore/heap/Heap.h:

Canonical link: <a href="https://commits.webkit.org/304020@main">https://commits.webkit.org/304020@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/42931cbcbd00f7cb69895a04f6663ddd211a8945

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows | Apple Internal |
| ----- | ---------------------- | ------- |  ----- |  --------- | ------ |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/134294 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/159/builds/6799 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/45515 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/141869 "Built successfully") | [  ~~🛠 win~~](https://ews-build.webkit.org/#/builders/59/builds/86330 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [![loading](https://user-images.githubusercontent.com/3098702/171232313-daa606f1-8fd6-4b0f-a20b-2cb93c43d19b.png) 🛠 ios-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/f2c6d2ed-bd8b-4d70-a499-64172ae81bd0/f12d4a15-97ad-474e-af4d-2e96e268c3f4) 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/155/builds/7381 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/156/builds/6662 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/102684 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [  ~~🧪 win-tests~~](https://ews-build.webkit.org/#/builders/59/builds/86330 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [![loading](https://user-images.githubusercontent.com/3098702/171232313-daa606f1-8fd6-4b0f-a20b-2cb93c43d19b.png) 🛠 mac-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/f2c6d2ed-bd8b-4d70-a499-64172ae81bd0/cb26c969-5e59-40f8-ab03-61269b098989) 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/137241 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/162/builds/5153 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/120372 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/83477 "Passed tests") | | ⏳ 🛠 vision-apple 
| | [  ~~🧪 ios-wk2-wpt~~](https://ews-build.webkit.org/#/builders/154/builds/5026 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-mac-debug~~](https://ews-build.webkit.org/#/builders/165/builds/2633 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [  ~~🛠 wpe-cairo-libwebrtc~~](https://ews-build.webkit.org/#/builders/166/builds/2343 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/126373 "Built successfully and passed tests") | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/13/builds/114227 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/38510 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/144521 "Built successfully") | | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/132810 "Built successfully and passed tests") | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/153/builds/6470 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/39088 "Passed tests") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/111072 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/160/builds/6554 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/161/builds/5443 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/111326 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/28239 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/164/builds/4855 "Passed tests") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/116641 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/60258 "Built successfully") | | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/158/builds/6523 "Built successfully") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/34852 "Passed tests") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/165767 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/157/builds/6357 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/70077 "Built successfully") | [  ~~🧪 jsc-armv7-tests~~](https://ews-build.webkit.org/#/builders/25/builds/43318 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [  ~~🛠 watch~~](https://ews-build.webkit.org/#/builders/163/builds/6594 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/152/builds/6470 "Built successfully") | | | | 
<!--EWS-Status-Bubble-End-->

dhecht reviewed (approved):
r=me but maybe the assert can be added using `vm->currentThreadIsHoldingAPILock()`

kmiller68 commented on Source/JavaScriptCore/heap/Heap.cpp:
```diff
@@ -1753,7 +1755,7 @@ NEVER_INLINE bool Heap::runEndPhase(GCConductor conn)
     // updateAllocationLimits() updates this for us.
     if (overCriticalMemoryThreshold())
-        sweepSynchronously();
+        m_needSynchronousSweep = true;
```
Nit: You could probably just remove this and ask if memory is critical in `shouldSweepSynchronously`

kmiller68 reviewed (commented):
r=me with nit too.

Commit:
GC sweeping should not be done in the End phase.
https://bugs.webkit.org/show_bug.cgi?id=303634
rdar://165681233

Reviewed by Dan Hecht and Keith Miller.

That's because the end phase may be run on the GC thread instead of the mutator thread.
Sweeping needs to be done on the mutator thread because AtomStringImpl destruction needs
to unregistered from the mutator's thread local AtomStringTable.  This can only be done
on the mutator thread.

303051@main introduced eager synchronous sweeping in the GC End phase if there's a critical
memory condition.  We'll fix this by changing this to do the critical memory check in
Heap::shouldSweepSynchronously() instead, and simply let Heap::finalize() do the sweeping
in the mutator thread instead (which it can already do).

Added a RELEASE_ASSERT in Heap::sweepSynchronously() to verify that the thread is holding
the API lock (as a proxy for identifying that it is the mutator thread).  This should
help catch this issue with existing tests to the extent possible.  The bug does rely on
a race condition that is not easily reproducible.

* Source/JavaScriptCore/heap/Heap.cpp:
(JSC::Heap::sweepSynchronously):
(JSC::Heap::runEndPhase):
(JSC::Heap::shouldSweepSynchronously):
* Source/JavaScriptCore/heap/Heap.h:

Canonical link: https://commits.webkit.org/304020@main
